staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -578,3 +604,57 @@ func (t CRDRESTOptionsGetter) GetRESTOptions(resource schema.GroupResource) (gen
 	}
 	return ret, nil
 }
+
+// defaultingRestOptionsGetter wraps a RestOptionGetter and sets a storage codec that does defaulting.
+type defaultingRestOptionsGetter struct {
+	getter    generic.RESTOptionsGetter
+	defaulter *validate.SchemaValidator
+}
+
+func (g defaultingRestOptionsGetter) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
+	opt, err := g.getter.GetRESTOptions(resource)
+	if err != nil {
+		return opt, err
+	}
+
+	cfg := *opt.StorageConfig // shallow copy
+	cfg.Codec = &defaultingCodec{opt.StorageConfig.Codec, defaultingDecoder{opt.StorageConfig.Codec, g.defaulter}}
+	opt.StorageConfig = &cfg
+
+	return opt, nil
+}
+
+// defaultingCodec does defaulting on decode
+type defaultingCodec struct {
+	runtime.Encoder
+	defaultingDecoder
+}
+
+// defaultingDecoder does defaulting on decode
+type defaultingDecoder struct {
+	runtime.Decoder
+	defaulter *validate.SchemaValidator
+}
+
+func (c defaultingDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	out, gvk, err := c.Decoder.Decode(data, defaults, into)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var u *unstructured.Unstructured
+	var ok bool
+	if u, ok = out.(*unstructured.Unstructured); !ok {
+		return nil, nil, fmt.Errorf("expected Unstructred in CR storage codec, got %T", out)
+	}
+
+	if c.defaulter != nil {
+		result := c.defaulter.Validate(u.UnstructuredContent())
```

Comment:
What's the perf impact of this?

Reply:
This depends a lot on the validation scheme. If it is huge, with many regex and especially quantors etc. it can be considerable. Non-native types are not for free for sure. The validation is `O(size of object + size of schema)` afaik. We have forbidden dangerous constructs like `UniqueItems´ already in the schema because they can easily lead to quadratic complexity (assuming Golang regex are O(n); also compare https://github.com/kubernetes/community/pull/708/files#diff-9679c445e428c00f0907fa6fe5c6d8e4R477).